src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +106,153 @@ def self.compile_ast(code, compile_options, boltlib_path)
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      # When we do not need to load bolt we assume there are no Bolt types to resolve in
+      # AST compilation.
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # Set the node_name_value setting directly
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+      # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
+      # not preserve hash order necessary for deserialization. The data strucutre is now stored
+      # in a list for moving data and the list is used to construct an ordered ruby hash.
+      def self.ordered_plan_vars(compile_options)
+        if compile_options['variables']['values'].is_a?(Array)
+          compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
+            acc[param_hash.keys.first] = param_hash.values.first
+          end
+        else
+          compile_options['variables']['values']
+        end
+      end
+      private_class_method :ordered_plan_vars
+
+      # Create a puppet settings context that can compile for projects
+      def self.with_puppet_settings_for_projects(overrides = {}, multithreaded)
+        unless multithreaded
+          # Create a copy of the existing settings object
+          original_settings = Puppet.settings
+        end
+
+        Dir.mktmpdir('bolt_compile') do |dir|
+          # This config sets all required settings to an empty dir
+          # where puppet won't load things like puppet.conf
+          puppet_config = {}
+          Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+            puppet_config[setting] = dir
+          end
```

Comment:
```suggestion
puppet_config = Puppet::Settings::REQUIRED_APP_SETTINGS.each_with_object({}) {|setting, acc| acc[setting] = dir}
```

Reply:
I'm fine being in the minority on this, but personally I find this pattern less readable than defining the var beforehand and not using each_with_object. I get why we would use this, and again I'm cool if everyone disagrees, I just don't think it reads as well.

Reply:
I agree with Sean tbh.

Reply:
I dont care too much. It is a method that does exactly a very common thing. But i also like list comprehensions in python. So take my opinion with a grain of salt.